### DISCUSSION TITLE
Programmable timers support in TuyaMCU

### DISCUSSION THREAD

**Discussion by OgnjenX** [arendst/Tasmota#12048] [Q&A] [answered]
I recently flashed and configured this device http://www.cn-hysen.com/productinfo/451126.html with TuyaMCU. Everything works fine except "programmable timers". I'm not sure how to configure them (and is it event possible currently), because of strange format and functionality (only 2 channels, one for workdays and one for weekends, and in these messages are stored and sent all data related to the timing at each day and target temperature). I would appreciate it if anyone has some knowledge and experience on something similar. Thanks in advance.

**Comment by barbudor** [comment, marked as answer]
If you are talking about the timers originally provided by the Tuya App, they were probably under the control of the TuyaMCU. So they probably should still be there and configurable through the device buttons and LCD.

For Tasmota timers, please refer to https://tasmota.github.io/docs/Timers/ and Web Gui "Configure" => "Timers"

**Reply by OgnjenX** [reply]
Thank you for your fast response. In my understanding link you provided is Tasmota native timers, I tried using them and they are working fine, but the firmware on the MCU of this device (and some similar devices) support their native timers, and my goal is to configure them. In my opinion, a good example of similar timers is shown in https://www.zigbee2mqtt.io/devices/TS0601_thermostat.html
(inside section "Controlling device specific features").

**Reply by barbudor** [reply]
As I said, the Tuya timers are under control of the TuyaMCU, so you can configure them by
- original buttons and LCD
- sending Tuya messages to the CPU with TuyaSend. Read https://tasmota.github.io/docs/Timers/

**Reply by OgnjenX** [reply]
Yes, you are right. Sorry for my misunderstanding at first. TuyaMCU timers are still configurable using original buttons and LCD, but I want to control them through MQTT. I understand it is theoretically possible by mapping proper fnId to dpId, and then controlling them using TuyaSend, as you said. But my problem is that if I configure them as timers (which are described in https://tasmota.github.io/docs/TuyaMCU/#timers ), I have not any control of them using TuyaSend. The only way that I get some control is if I configure them as dimmers (but then they are not changing meaningfully at all, and also they make an impact on other channels changing, which makes me just more confused).

**Reply by Jason2866** [reply]
Tuya has no fixed standards. Every manufactor can change. To get Tuya based functions running with Tasmota is not always easy / possible.

**Reply by OgnjenX** [reply]
I figured out that channels, through which messages are sent, have DpType  0. I tried all kinds of TuyaSend commands (String - TuyaSend3 would be the best guess) but I can't get proper behavior.
Does DpType 0 mean that type of that channel is not fully supported by Tasmota yet, and could not be controlled properly, or it means something else (also if there is a known solution)?
Thanks in advance.
